src/main/java/codechicken/nei/PanelWidget.java

```diff
@@ -0,0 +1,375 @@
+        }
+
+        if (handleGUIContainerClick(mouseX, mouseY, button)) {
+            return true;
+        }
+
+        final GuiContainer gui = NEIClientUtils.getGuiContainer();
+        if (mouseX < gui.guiLeft || mouseY < gui.guiTop || mouseX >= gui.guiLeft + gui.xSize || mouseY >= gui.guiTop + gui.ySize) {
+            draggedStack = null;
+        }
+
+        return true;
+    }
+
+    protected boolean handleGUIContainerClick(int mouseX, int mouseY, int button)
+    {
+        final GuiContainer gui = NEIClientUtils.getGuiContainer();
+        boolean handled = false;
+
+        try {
+            GuiInfo.readLock.lock();
+            for (INEIGuiHandler handler : GuiInfo.guiHandlers) {
+                if (handler.handleDragNDrop(gui, mouseX, mouseY, draggedStack, button)) {
+                    handled = true;
+                    break;
+                }
+            }
+        } finally {
+            GuiInfo.readLock.unlock();
+        }
+
+        if (draggedStack.stackSize == 0) {
+            draggedStack = null;
+        }
+
+        return handled;
+    }
+
+    @Override
+    public void mouseUp(int mousex, int mousey, int button)
+    {
+        ItemPanelSlot hoverSlot = getSlotMouseOver(mousex, mousey);
+
+        if (hoverSlot != null && hoverSlot.slotIndex == mouseDownSlot && draggedStack == null) {
+            ItemStack item = hoverSlot.item;
+
+            if (NEIController.manager.window instanceof GuiRecipe || !NEIClientConfig.canCheatItem(item)) {
+
+                if (button == 0) {
+                    GuiCraftingRecipe.openRecipeGui("item", item);
+                } else if (button == 1) {
+                    GuiUsageRecipe.openRecipeGui("item", item);
+                }
+
+                mouseDownSlot = -1;
+                return;
+            }
+
+            NEIClientUtils.cheatItem(getDraggedStackWithQuantity(hoverSlot.slotIndex), button, -1);
+        }
+
+        mouseDownSlot = -1;
+    }
+
+    @Override
+    public boolean onMouseWheel(int i, int mousex, int mousey)
+    {
+        if (!contains(mousex, mousey))
+            return false;
+
+        grid.shiftPage(-i);
+        return true;
+    }
+
+    @Override
+    public boolean handleKeyPress(int keyID, char keyChar)
+    {
+
+        if (keyID == NEIClientConfig.getKeyBinding("gui.next")) {
+            grid.shiftPage(1);
+            return true;
+        }
+
+        if (keyID == NEIClientConfig.getKeyBinding("gui.prev")) {
+            grid.shiftPage(-1);
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public ItemStack getStackMouseOver(int mousex, int mousey)
+    {
+        ItemPanelSlot slot = getSlotMouseOver(mousex, mousey);
+        return slot == null ? null : slot.item;
+    }
+
+    public ItemPanelSlot getSlotMouseOver(int mousex, int mousey)
+    {
+        return grid.getSlotMouseOver(mousex, mousey);
+    }
+
+    public int getPage() {
+        return grid.page + 1;
```

Comment:
I'd return 0-based indices here, and only convert to 1-based where it's displayed in the UI

Reply:
This is old method (moved from another class) and i save backward compatibility. I'm not sure if other mods don't use it